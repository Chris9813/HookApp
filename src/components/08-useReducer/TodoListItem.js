import React from 'react'

export const TodoListItem = ({todo, index, handleDelete, handleToggle}) => {
    return (
        <div>
            <p
            className = {`${todo.done && "complete"}`}
            onClick = {() => handleToggle(todo.id)}
            >{index+1}. {todo.desc}</p>
            <button 
            key = {todo.id}
            className = "btn btn-danger"
            onClick = {() => handleDelete(todo.id)}
            > Borrar</button>
        </div>
    )
}
