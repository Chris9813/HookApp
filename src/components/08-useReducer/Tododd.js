import React from 'react'
import { useForm } from '../../hooks/useForm'



export const Tododd = ({handleAddTodo}) => {

    const [{description}, handleInnputChange, reset] = useForm({
        description: ""
    })

    const handleSubmit = (e) => {
        e.preventDefault();
        if (description.trim().length <= 1){
            return;
        }
        const newTodo = {
        id: new Date().getTime(),
        desc: description,
        done: false,
        }
        handleAddTodo(newTodo);
        reset()
        }

    return (
        <>
                <h4>Agregar TODO</h4>
            <hr/>
            <form onSubmit = {handleSubmit}>
            <input
            placeholder = "Agregar elemento"
            type ="text"
            autoComplete = "off"
            className = "form-control"
            name = "description"
            value = {description}
            onChange = {handleInnputChange}
            ></input>
            <div className = "d-grid gap-2">
            <button
            type = "submit"
            className = "btn btn-outline-primary mt-3"
            >Enviar</button>
            </div>
            
        </form>
        </>
    )
}
