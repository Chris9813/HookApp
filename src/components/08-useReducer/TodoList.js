import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({todos, handleDelete, handleToggle}) => {

    return (
        <ul className ="list-group list-group-flush">
                {
                todos.map((todo, i) =>(
                    <li
                    key = {todo.id}>
                        <TodoListItem 
                        todo = {todo}
                        index = {i}
                        handleDelete = {handleDelete}
                        handleToggle = {handleToggle}
                        />
                    </li>
                ))
                }
        </ul>)
}

/* 
todo = {todo}
            index = {index}
            handleDelete = {handleDelete}
            handleToggle = {handleToggle}
            />
*/
