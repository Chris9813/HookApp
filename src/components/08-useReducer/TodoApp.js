import React, { useEffect, useReducer } from 'react'
import { useForm } from '../../hooks/useForm'
import { todoReducer } from './todoReducer'
import "./styles.css"
import { TodoList } from './TodoList'
import { Tododd } from './Tododd'




export const TodoApp = () => {

    const init = () => {
        return JSON.parse(localStorage.getItem("todos")) ||  [] 
    }

    const [todos, dispatch] = useReducer(todoReducer,  [], init )
    

    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todos))
    }, [todos])

    const handleDelete = (todoId) => {
        
        const action = {
            type: "delete",
            payload: todoId,
        }
        dispatch(action)
    }
    
    const handleToggle = (todoId) => {
    dispatch({
        type: "toggle",
        payload: todoId
    })
    }

    const handleAddTodo = (newTodo) => {

    dispatch({
        type: "add",
        payload: newTodo,
    })
    }

    

    return (
        <div>
        <h1>Todos App: {todos.length}</h1>
        <hr/>
        <div className = "row">
        <div className = "col-7">
        
        < TodoList 
        todos = {todos}
        handleDelete = {handleDelete}
        handleToggle = {handleToggle}
        />

        </div>
        
        <div className = "col-5 text-center">
            <Tododd 
            handleAddTodo = {handleAddTodo}
            />

        </div>
        
        </div>
        
        </div>
    )

}



/*
const init = () => {

    return JSON.parse(localStorage.getItem("todos")) ||  [] 

}
export const TodoApp = () => {
    const [todos, dispatch] = useReducer(todoReducer, [], init)
    const [{description}, handleInnputChange, reset] = useForm({
        description: ""
    })

    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todos))
    }, [todos])

    console.log(description);
    

    const handleSubmit = (e) => {
        e.preventDefault();
        if(description.trim().length < 1){
            return;
        }
        const newTodo = {
        id: new Date().getTime(),
        desc: description,
        done: false
        } 
        const action = {
            type: "add",
            payload: newTodo
        }
        dispatch( action );
        reset();
    }

    return (
        <div>
            <h1>TodoApp ({todos.length})</h1>
            <hr/>
            <div className ="row">
                <div className="col-7">
                <ul className ="list-group list-group-flush">
                {
                    todos.map((todo, i) => (
                    <li
                    key = {todo.id}
                    className = "list-group-item"
                    >
                    <p
                    className = "text-center"
                    >{i + 1}. {todo.desc}</p> 
                    
                    <button
                    className = "btn btn-danger"
                    >
                        Borrar
                    </button>
                    </li>
                    ))
                }
            </ul>
            </div>
            <div className ="col-5">
                    <h4>Agregar Todo</h4>
                    <hr/>
                    <form
                    onSubmit = {handleSubmit}
                    >
                        <input
                        type= "text"
                        name = "description"
                        placeholder = "Aprender.."
                        autoComplete = "off"
                        className ="form-control"
                        onChange = {handleInnputChange}
                        value = {description}
                        >
                        </input>
                        <button
                        type = "submit"
                        className= "btn btn-outline-primary mt-2 btn-block"
                        >
                        Agregar
                        </button>
                    </form>
            </div>
            </div>
            
        </div>
    )
}
*/