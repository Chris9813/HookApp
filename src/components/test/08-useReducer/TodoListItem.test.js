import { shallow } from 'enzyme'
import React from 'react'
import { TodoListItem } from '../../08-useReducer/TodoListItem'
import { demoTodos } from '../fixtures/demoTodos'


describe("Pruebas en TodoLisItem", () => {

    const handleDelete = jest.fn()
    const handleToggle = jest.fn()

    const wraper = shallow(<TodoListItem 
    todo = {demoTodos[0]}
    index = {0}
    handleDelete = {handleDelete}
    handleToggle = {handleToggle}
    />)

    test('debe mostrarse correctamnete ', () => {
        expect(wraper).toMatchSnapshot();
    })

    test('debe llamar a la funcion handleDelete ', () => {
        wraper.find("button").simulate("click")
        expect(handleDelete).toHaveBeenCalledWith(1);
    })
    
    test('debe llamar a la funcion handleToggle ', () => {
        wraper.find("p").simulate("click")
        expect(handleToggle).toHaveBeenCalledWith(1);
    })

    test('debe llamar mostrar correctamente el parrafo ', () => {
        const p = wraper.find("p")
        expect(p.text().trim()).toBe(`1. ${demoTodos[0].desc}`)
    })

    test('debe tener a case complete si esta completo', () => {
        
        const todo = demoTodos[0]
        todo.done = true

        const wraper = shallow(<TodoListItem 
            todo = {todo}
            />)
        expect( wraper.find("p").hasClass("complete") ).toBe(true)
    })
    
    
    
    
})