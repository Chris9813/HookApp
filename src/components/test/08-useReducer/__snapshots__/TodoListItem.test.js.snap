// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoLisItem debe mostrarse correctamnete  1`] = `
<div>
  <p
    className="false"
    onClick={[Function]}
  >
    1
    . 
    aprender Mongo
  </p>
  <button
    className="btn btn-danger"
    key="1"
    onClick={[Function]}
  >
     Borrar
  </button>
</div>
`;
